File: Grind75/py/TwoSum1_2_3.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i+1,len(nums)):
                if nums[i] + nums[j] == target:
                    return [i,j]


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # {
        #     7:0,
        #     2:1,
        #     -3:2,
        #     -6:3
        # }
        dict ={}
        for i, n in enumerate(nums):
            if nums[i] in dict:
                return dict[n], i # dict[n] + i = target
            else: #dict에 넣는 부분
                dict[target-n] = i
        # print(dict)



class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums = list(enumerate(nums))
        nums.sort(key=lambda x:x[1])
        left,right = 0,len(nums)-1
        while left < right:
            if nums[left][1] + nums[right][1] == target:
                return [nums[left][0], nums[right][0]]
            elif nums[left][1] + nums[right][1] < target:
                left += 1
            else:
                right -=1
            

File: 2024/Grind75/TwoSum1_2_3.java
package Grind75;
import java.util.HashMap;
import java.util.Map;

public class TwoSum1_2_3 {

    // public int[] twoSum(int[] nums, int target) {
    //     int n = nums.length;
    //     for (int i = 0; i < n - 1; i++) {
    //         for (int j = i + 1; j < n; j++) {
    //             if (nums[i] + nums[j] == target) {
    //                 return new int[]{i, j};
    //             }
    //         }
    //     }
    //     return new int[]{}; // No solution found
    // }

    // public int[] twoSum(int[] nums, int target){
    //     Map<Integer, Integer> numMap = new HashMap<>();
    //     int n = nums.length;

    //     for(int i = 0; i < n; i++){
    //         numMap.put(nums[i],i);
    //     }

    //     for (int i = 0; i < n; i++) {
    //         int complement = target - nums[i];
    //         if(numMap.containsKey(complement) && numMap.get(complement) !=i){
    //             return new int[]{i, numMap.get(complement)};
    //         }
    //     }
    //     throw new IllegalArgumentException("No two sum solution");
    // }

    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> numMap = new HashMap<>();
        int n = nums.length;

        for (int i = 0; i < n; i++) {
            int complement = target - nums[i];
            if (numMap.containsKey(complement)) {
                return new int[]{numMap.get(complement), i};
            }
            numMap.put(nums[i], i);
        }

        return new int[]{}; // No solution found
    }

        public static void main(String[] args) {
            TwoSum1_2_3 solution = new TwoSum1_2_3();
            int[] nums = {2,7,11,15};
            int target = 9;
            int[] result = solution.twoSum(nums, target);

            for(int i : result){
                System.out.print(i + " ");
            }
        }
}
